File: app/src/main/java/com/sergeystets/vladex/activity/SignInActivity.java
package com.sergeystets.vladex.activity;

import android.content.ComponentName;
import android.content.res.AssetManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.customtabs.CustomTabsClient;
import android.support.customtabs.CustomTabsIntent;
import android.support.customtabs.CustomTabsServiceConnection;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import com.sergeystets.vladex.R;
import com.sergeystets.vladex.api.VladexApi;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

import static com.sergeystets.vladex.util.Utils.TAG;

public class SignInActivity extends AppCompatActivity {

    private final Properties properties = new Properties();

    @Inject
    VladexApi vladexApi;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.sign_in_activity);

        final AssetManager assetManager = this.getAssets();
        String authUrl;
        try (final InputStream inputStream = assetManager.open("application.properties")) {
            properties.load(inputStream);
            authUrl = properties.getProperty("app.vladex.auth.url");
        } catch (IOException e) {
            Log.e(TAG, "Failed to load application.properties", e);
            return;
        }

        final SignInActivity context = this;
        final Button signIn = findViewById(R.id.sign_in);
        final Uri url = Uri.parse(authUrl + "/oauth/authorize"
                + "?response_type=token"
                + "&client_id=vladex-mobile"
                + "&scope=API"
                + "&redirect_uri=app://vladex-messenger"
                + "&grant_type=implicit");

        signIn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                final CustomTabsServiceConnection connection = new CustomTabsServiceConnection() {
                    @Override
                    public void onCustomTabsServiceConnected(ComponentName componentName, CustomTabsClient client) {
                        final CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
                        final CustomTabsIntent intent = builder.build();
                        client.warmup(0L); // This prevents back grounding after redirection
                        intent.launchUrl(context, url);
                    }

                    @Override
                    public void onServiceDisconnected(ComponentName name) {
                    }
                };
                CustomTabsClient.bindCustomTabsService(context, "com.android.chrome", connection);
            }
        });
    }
}

File: app/src/main/java/com/sergeystets/vladex/SignInActivity.kt
package com.sergeystets.vladex

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.auth.api.signin.GoogleSignInOptions.DEFAULT_SIGN_IN
import com.google.android.gms.common.api.ApiException

class SignInActivity : Activity(), View.OnClickListener {

    private var signInClient: GoogleSignInClient? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_in)

        findViewById<View>(R.id.sign_in_button).setOnClickListener(this)

        setupGoogleSignInClient()
    }

    private fun setupGoogleSignInClient() {
        val gso = GoogleSignInOptions.Builder(DEFAULT_SIGN_IN)
                .requestEmail()
                .requestProfile()
                .build()

        signInClient = GoogleSignIn.getClient(this, gso)
    }

    override fun onClick(view: View) {
        if (view.id == R.id.sign_in_button) {
            signIn()
        }
    }

    private fun signIn() {
        Log.i(TAG, "signIn in...")
        val signInIntent = signInClient!!.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == RC_SIGN_IN) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            try {
                val account = task.getResult(ApiException::class.java)
                goToHomeScreen(account)
            } catch (e: ApiException) {
                Log.w(TAG, "failed to sign in", e)
            }
        }
    }

    private fun goToHomeScreen(account: GoogleSignInAccount?) {
        if (account != null) {
            val homeScreen = Intent(this, HomeScreenActivity::class.java)
            startActivity(homeScreen)
            Log.i(TAG, "user with email: ${account.email}' successfully logged in")
        }
    }

    companion object {
        private const val TAG = "SignInActivity"
        private const val RC_SIGN_IN = 9001
    }
}
